File: leetcode/golang/72_edit_distance.go
func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func minDistance(word1 string, word2 string) int {
    n1, n2 := len(word1), len(word2)
    dp := make([][]int, n1+1)
    for i:= 0; i <= n1; i++ {
        dp[i] = make([]int, n2+1)
        for j := 0; j <= n2; j++ {
            if i == 0 {
                dp[i][j] = j
            } else if j == 0 {
                dp[i][j] = i
            } else {
                dp[i][j] = 600
            }
        }
    }
    for i := 1; i <= n1; i++ {
        for j := 1; j <= n2; j++ {
            if word1[i-1] == word2[j-1] {
                dp[i][j] = dp[i-1][j-1]
            } else {
                dp[i][j] = min(min(dp[i-1][j-1], dp[i][j-1]), dp[i-1][j]) + 1
            }
        }
    }
    return dp[n1][n2]
}

File: leetcode/c++/72_edit_distance.cpp
// https://leetcode.cn/problems/edit-distance

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minDistance(string word1, string word2) {
        int n1 = word1.length(), n2 = word2.length();
        vector<vector<int>> dp(n1+1, vector<int>(n2+1, 600));
        for (int i = 0; i <= n1; i++) 
            dp[i][0] = i;
        for (int i = 0; i <= n2; i++) 
            dp[0][i] = i;
        for (int i = 1; i <= n1; i++) {
            for (int j = 1; j <= n2; j++) {
                if (word1[i-1] == word2[j-1])
                    dp[i][j] = dp[i-1][j-1];
                else 
                    dp[i][j] = min(min(dp[i-1][j], dp[i][j-1]), dp[i-1][j-1]) + 1;
            }
        }
        return dp[n1][n2];
    }
};

// 记得每次操作要+1+1+1！！！
// 时间复杂度为O(nm)，空间复杂度为O(nm)
// dp[i-1][j]：删除word1[i-1]
// dp[i][j-1]：增加word1[i-1]
// dp[i-1][j-1]：替换word1[i-1] = word[j-1]